oeps/oep-0036-microservice-routing-layer.rst

```diff
@@ -0,0 +1,99 @@
+===================================
+OEP-36: Microservice Routing Layer
+===================================
+
+.. list-table::
+
+  * - OEP
+    - :doc:`oep-0036-microservice-routing-layer`
+  * - Title
+    - Microservice Routing Layer
+  * - Last-Modified
+    - 2019-04-25
+  * - Author
+    - Fred Smith <derf@edx.org>
+  * - Arbiter
+    - 
+  * - Status
+    - Provisional
+  * - Type
+    - Architecture
+  * - Created
+    - 2019-04-25
+
+Abstract
+========
+
+Proposes a system with which users can interact with microservice applications
+using a layer 7 routing framework. 
+
+Motivation
+==========
+
+Microservices are currently implemented using layer 4 (DNS) based routing. Each
+microservice occupies a unique hostname, with dedicated ELB or CDN 
+infrastructure to support routing user traffic to the service. This pattern
+is simple to monitor, deploy, and manage, but slow and expensive to build,
+and regularly causes problems with cross service user information sharing. In
+Addition, modifying or splitting a monolithic service into microservices resuls
+in confusing URL changes and costly migrations.
+
+A layer 7 routing layer would allow us to separate the user facing URL or API
+endpoint from the backend implementation. It would allow microservices to exist
+that only manage a small number of endpoints, without requring developers to 
+build out load balancing or CDN infrastructure. 
+
+We are currently using a layer 7 routing framework for prospectus, allowing
+us to replace drupal generated pages with gatsby generated pages quickly and
+seamlessly.
+
+This change will allow developers, openedx operators and resellers to deploy
+new openedx sites quicker and easier than they currently can.  The router
+will provide a map for people to understand which services they need to deploy.
+
+
+Specification
```

Comment:
This might be implied because it's in an OEP, but...

Since this solution is replacing a vendor specific solution we have today, it would be good to note that this would be a separate application that is part of the Open edX stack and therefore it can run independently both in production and locally in a devstack. 

Partially related to this: In the MFE world, will this only act as a router to a separate service, or will this also be able to host static files? Internally we have a solution that's DNS + static webserver for MFEs. Will this handle both of those use cases, or will MFE's static webserver be a separate OEP / service?

Reply:
I think the question of where static content is hosted depends on the tech choice - some layer 7 router implementations (eg. traefik, ambassador) don't support hosting static content and require you to run a static content server like nginx